Code/Framework/AzCore/AzCore/DOM/DomPath.cpp

```diff
@@ -0,0 +1,478 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Casting/numeric_cast.h>
+#include <AzCore/DOM/DomPath.h>
+#include <AzCore/std/string/fixed_string.h>
+#include <AzCore/Console/ConsoleTypeHelpers.h>
+
+namespace AZ::Dom
+{
+    PathEntry::PathEntry(size_t value)
+        : m_value(value)
+    {
+    }
+
+    PathEntry::PathEntry(AZ::Name value)
+        : m_value(AZStd::move(value))
+    {
+    }
+
+    PathEntry::PathEntry(AZStd::string_view value)
+        : m_value(AZ::Name(value))
+    {
+    }
+
+    PathEntry& PathEntry::operator=(size_t value)
+    {
+        m_value = value;
+        return *this;
+    }
+
+    PathEntry& PathEntry::operator=(AZ::Name value)
+    {
+        m_value = AZStd::move(value);
+        return *this;
+    }
+
+    PathEntry& PathEntry::operator=(AZStd::string_view value)
+    {
+        m_value = AZ::Name(value);
+        return *this;
+    }
+
+    bool PathEntry::operator==(const PathEntry& other) const
+    {
+        return m_value == other.m_value;
+    }
+
+    bool PathEntry::operator==(size_t value) const
+    {
+        return IsIndex() && GetIndex() == value;
```

Comment:
Similar to how the change to using `get_if` worked out elsewhere, it will be more performant if this is replaced with 
```
const size_t* index = AZStd::get_if<size_t>(&m_value);
return result != nullptr ? ((*result) == value) : false;
```
Similar for the `operator==`s and `operator!=`s below.

Reply:
:thumbsup: Merging this for now but will follow up with the remaining suggestions for the Patch PR (which will also give me another set of benchmarks for a bit more context).